pg_lakehouse/src/datafusion/session.rs

```diff
@@ -29,31 +31,134 @@ impl Session {
             &'a SessionContext,
         ) -> Pin<Box<dyn Future<Output = Result<R, ContextError>> + 'a>>,
     {
-        let mut lock = task::block_on(SESSION_CACHE.lock());
+        let mut lock = task::block_on(SESSION_CACHE.write());
 
         let context = match lock.entry(SESSION_ID.to_string()) {
             Occupied(entry) => entry.into_mut(),
-            Vacant(entry) => {
-                // Set current timezone
-                let mut session_config = SessionConfig::from_env()?.with_information_schema(true);
-                let session_timezone = unsafe {
-                    CStr::from_ptr(pg_sys::pg_get_timezone_name(pg_sys::session_timezone))
-                        .to_str()
-                        .unwrap_or_else(|err| panic!("{:?}", err))
-                };
-                session_config.options_mut().execution.time_zone =
-                    Some(session_timezone.to_string());
-
-                // Create a new context
-                let rn_config = RuntimeConfig::new();
-                let runtime_env = RuntimeEnv::new(rn_config)?;
-                let context =
-                    SessionContext::new_with_config_rt(session_config, Arc::new(runtime_env));
-
-                entry.insert(context)
-            }
+            Vacant(entry) => entry.insert(Self::init()?),
         };
 
         task::block_on(f(context))
     }
+
+    pub fn with_session_context_read<F, R>(f: F) -> Result<R, ContextError>
+    where
+        F: for<'a> FnOnce(
+            &'a SessionContext,
+        ) -> Pin<Box<dyn Future<Output = Result<R, ContextError>> + 'a>>,
+    {
+        let lock = task::block_on(SESSION_CACHE.read());
+
+        let context = match lock.get(SESSION_ID) {
+            Some(entry) => entry,
+            None => return Err(ContextError::ContextNotFound),
+        };
+
+        task::block_on(f(context))
+    }
+
+    pub fn with_catalog<F, R>(f: F) -> Result<R, ContextError>
+    where
+        F: for<'a> FnOnce(
+            &'a LakehouseCatalog,
+        ) -> Pin<Box<dyn Future<Output = Result<R, ContextError>> + 'a>>,
+    {
+        let mut lock = task::block_on(SESSION_CACHE.write());
+
+        let context = match lock.entry(SESSION_ID.to_string()) {
+            Occupied(entry) => entry.into_mut(),
+            Vacant(entry) => entry.insert(Self::init()?),
+        };
+
+        let catalog_provider = context.catalog(&Self::catalog_name()?).ok_or(
+            ContextError::CatalogProviderNotFound(Self::catalog_name()?.to_string()),
+        )?;
+
+        let downcast_catalog = catalog_provider
+            .as_any()
+            .downcast_ref::<LakehouseCatalog>()
+            .ok_or(ContextError::CatalogNotFound(
+                Self::catalog_name()?.to_string(),
+            ))?;
+
+        task::block_on(f(downcast_catalog))
+    }
+
+    pub fn with_schema_provider<F, R>(schema_name: &str, f: F) -> Result<R, ContextError>
```

Comment:
I'm not sure I fully understand the reasoning for the `with_` pattern. It seems like we're just using it to get access to a statically-owned resource.

Why couldn't `Session::schema_provider` just return a `&LakehouseSchemaProvider`? That would save the caller from having to create a closure (and all the lifetime annoyances that come with that).

I'm open to there being a good reason why you're doubling-down on this pattern (maybe it actually makes lifetime management easier, somehow?), but it's not clear at first glance.

Reply:
The reason I introduced this pattern was so Rust knows to release the lock at the end of the closure. If it releases a reference, how would it know to release the lock? Would that even be allowed?

Reply:
Do you need to keep the lock on the resource all the way through the call to `f`? If not, I think you should just clone `context` and return it to the caller. It doesn't seem like you want to keep the lock the whole time.

That being said, I'm aware that we're in a single threaded context and this doesn't matter that much. I'm only bringing it up because you're expanding the use of the pattern here, and want to make sure it makes sense for us. Your reasoning is good, if you've tried a simpler method and found this works best, then I'm happy with it.